sync/src/types.rs

```diff
@@ -1112,17 +1120,21 @@ impl ActiveChain {
     }
 
     pub fn get_ancestor(&self, base: &Byte32, number: BlockNumber) -> Option<core::HeaderView> {
-        // shortcut to return a ancestor block
-        if self.snapshot.is_main_chain(&base) {
-            return self.snapshot.get_block_hash(number).and_then(|hash| {
-                self.shared
-                    .get_header_view(&hash)
-                    .map(HeaderView::into_inner)
-            });
-        }
-        self.shared
-            .get_header_view(base)?
-            .get_ancestor(number, |hash| self.shared.get_header_view(hash))
+        let tip_number = self.tip_number();
+        self.shared.get_header_view(base)?.get_ancestor(
+            number,
+            |hash| self.shared.get_header_view(hash),
+            |current| {
+                // shortcut to return a ancestor block
+                if current.number() <= tip_number && self.snapshot().is_main_chain(&current.hash())
```

Comment:
@zhangsoledad Should we hold the snapshot instance locally to make sure `get_block_hash` will use the same snapshot?

Reply:
rebase will resolve this.

Reply:
Checked the code. The Arc is cloned when creating the `active_chain`, and will not change in the `active_chain` instance.